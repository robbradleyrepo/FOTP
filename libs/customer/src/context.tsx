import React, {
  createContext,
  FC,
  useCallback,
  useContext,
  useEffect,
  useState,
} from "react";

import { getTraitsFromCookie, setTraitsCookie, Traits } from "./traits";
import { Customer } from "./types";

interface CustomerContextProps {
  customer: Customer | null;
  loading: boolean;
  setTraits: (traits: Traits) => void;
}

const CustomerContext = createContext<CustomerContextProps | null>(null);

export const CustomerProvider: FC = ({ children }) => {
  const [value, setValue] = useState<Omit<CustomerContextProps, "setTraits">>({
    customer: null,
    loading: true,
  });

  // Load the customer data after the first render to ensure consistency
  // between client and server
  useEffect(() => {
    const customer = getTraitsFromCookie();

    setValue({ customer, loading: false });
  }, []);

  const setTraits = useCallback(
    (traits: Traits) => {
      setTraitsCookie(traits);
      setValue({ ...value, customer: traits });
    },
    [value]
  );

  return (
    <CustomerContext.Provider value={{ ...value, setTraits }}>
      {children}
    </CustomerContext.Provider>
  );
};

export const useCustomer = () => {
  const context = useContext(CustomerContext);

  if (!context) {
    throw new Error("`useCustomer` must be used inside a `CustomerProvider`");
  }

  return context;
};
