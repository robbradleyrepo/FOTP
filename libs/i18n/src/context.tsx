import React, { createContext, FC, useContext } from "react";
import { I18nextProvider, useTranslation } from "react-i18next";

import { getConfig, getI18nInstance } from "./config";
import { Locale } from "./types";

const LocaleContext = createContext<Locale | null>(null);

interface LocaleProviderProps {
  locale: Locale;
}

export const LocaleProvider: FC<LocaleProviderProps> = ({
  children,
  locale,
}) => {
  const i18n = getI18nInstance();

  // Make sure the correct language is set before rendering
  if (locale.langtag !== i18n.language) {
    i18n.changeLanguage(locale.langtag);
  }

  return (
    <LocaleContext.Provider value={locale}>
      <I18nextProvider i18n={i18n}>{children}</I18nextProvider>
    </LocaleContext.Provider>
  );
};

export const useLocale = () => {
  const { fallbackLocale } = getConfig();

  const locale = useContext(LocaleContext) ?? fallbackLocale;

  return {
    ...useTranslation(),
    locale,
  };
};
