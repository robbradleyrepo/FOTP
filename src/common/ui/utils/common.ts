import {
  ResponsiveCSSObject,
  ResponsiveCSSObjectOrValue,
  Style,
  ThemeEnhanced,
} from "./types";

export const evaluate = (
  style: Style,
  theme: ThemeEnhanced
): ResponsiveCSSObject => (typeof style === "function" ? style(theme) : style);

export const isObject = (
  target: ResponsiveCSSObjectOrValue
): target is ResponsiveCSSObject =>
  Object.prototype.toString.call(target) === "[object Object]";

export const merge = (
  accum: ResponsiveCSSObject,
  ...inputs: ResponsiveCSSObject[]
): ResponsiveCSSObject => {
  if (!inputs.length) return accum;

  const input = inputs.shift();
  let output = accum;

  for (const key in input) {
    const left = accum[key];
    const right = input[key];

    output = {
      ...output,
      [key]: isObject(left) && isObject(right) ? merge(left, right) : right,
    };
  }

  return merge(output, ...inputs);
};
