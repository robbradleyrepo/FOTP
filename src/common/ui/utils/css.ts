import { CSSValue, ResponsiveCSSObject, ResponsiveCSSValue } from "./types";

export const mx = (value: ResponsiveCSSValue): ResponsiveCSSObject => ({
  marginLeft: value,
  marginRight: value,
});

export const my = (value: ResponsiveCSSValue): ResponsiveCSSObject => ({
  marginBottom: value,
  marginTop: value,
});

export const negate = (value: CSSValue) => {
  if (typeof value === "string") {
    return `-${value}`;
  } else if (typeof value === "number") {
    return -value;
  }

  return value;
};

export const percentage = (value: number) => value * 100 + "%";

export const px = (value: ResponsiveCSSValue): ResponsiveCSSObject => ({
  paddingLeft: value,
  paddingRight: value,
});
export const py = (value: ResponsiveCSSValue): ResponsiveCSSObject => ({
  paddingBottom: value,
  paddingTop: value,
});

export const size = (value: ResponsiveCSSValue): ResponsiveCSSObject => ({
  height: value,
  width: value,
});
