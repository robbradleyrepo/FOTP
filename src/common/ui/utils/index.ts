export * from "./composer";
export * from "./css";
export * from "./helpers";
export * from "./provider";
export * from "./styles";
export * from "./types";
