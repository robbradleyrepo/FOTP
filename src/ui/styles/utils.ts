import { color } from "./variables";

interface DataUriFromPathParams {
  fill?: string;
  path: string;
  viewBox?: string;
  [k: string]: number | string | undefined;
}

const dataUriFromPath = ({
  fill = color.text.dark.base,
  path,
  viewBox = "0 0 100 100",
  ...rest
}: DataUriFromPathParams) => {
  const attributes = Object.entries(rest)
    .map(([key, value]) => ` ${key}="${value}"`)
    .join("");

  return `data:image/svg+xml;charset=utf8,${encodeURIComponent(`<svg viewBox="${viewBox}" xmlns="http://www.w3.org/2000/svg"${attributes}><path d="${path}" fill="${fill}" /></svg>
`)}`;
};

export { dataUriFromPath };
