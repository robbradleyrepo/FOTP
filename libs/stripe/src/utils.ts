import { Countries } from "@sss/ecommerce";
import {
  AddressInput,
  isPartialAddressInput,
  PartialAddressInput,
} from "@sss/ecommerce/address";
import { ShippingRate } from "@sss/ecommerce/checkout";
import { Money, moneyFns } from "@sss/ecommerce/common";
import {
  Address,
  PaymentRequestShippingAddress,
  PaymentRequestShippingOption,
} from "@stripe/stripe-js";
import { parseFullName as baseParseFullName } from "parse-full-name";

// The `parse-full-name` library doesn't handle middle initials that are also
// conjunctions - see https://github.com/RateGravity/parse-full-name/issues/7
const parseFullName = (fullName: string) => {
  // Based on the `conjunctionList` array that causes this issue
  const conjunctionRe = /\s(&|and|et|e|of|the|und|y)\s/gi;
  const result = baseParseFullName(fullName);

  return !result.first || result.error?.length
    ? // Try again, but remove any conjunctions. This might mangle a few
      // legitimate names, but we need a first name to fulfil orders
      baseParseFullName(fullName.replace(conjunctionRe, " "))
    : result;
};

// We're using || instead of nullish as Stripe does not allow empty strings.
const sanitise = (value?: string) => value?.trim() || undefined;

export const fromPaymentRequestShippingAddress = <T extends boolean = false>(
  {
    addressLine,
    dependentLocality,
    city,
    country,
    phone,
    postalCode,
    recipient,
    region,
  }: PaymentRequestShippingAddress,
  partial?: T
) => {
  let name: Pick<AddressInput, "firstName" | "lastName"> | null = null;

  if (recipient) {
    const { error, first, last } = parseFullName(recipient);

    name =
      !error?.length && !!first && !!last
        ? { firstName: first, lastName: last }
        : { firstName: "", lastName: recipient }; // Fall back to the unparsed name
  }

  const address: AddressInput | PartialAddressInput = {
    ...name,
    address1: addressLine?.[0],
    address2: addressLine?.[1] ?? dependentLocality ?? "",
    city: city ?? "",
    company: "",
    countryCode: country as keyof typeof Countries,
    phone: phone ?? "",
    province: region ?? "",
    zip: postalCode ?? "",
  };

  if (!partial && isPartialAddressInput(address)) {
    throw new Error("Missing address fields");
  }

  // Cast to conditional type as TypeScript can't infer it
  return address as T extends true ? PartialAddressInput : AddressInput;
};

export const fromBillingDetails = ({
  address: { city, country, line1, line2, postal_code, state },
  name,
  phone,
}: {
  address: Address;
  name: string;
  phone: string;
}): AddressInput => {
  if (!country || !line1 || !name || !postal_code || !state) {
    throw new Error("Missing data in billing details");
  }

  let firstName = "";
  let lastName = name;

  const { error, first, last } = parseFullName(name);

  if (!error?.length && !!first && !!last) {
    firstName = first;
    lastName = last;
  }

  return {
    address1: line1,
    address2: line2 ?? "",
    city: city ?? "",
    company: "",
    countryCode: country as keyof typeof Countries,
    firstName,
    lastName,
    phone: phone,
    province: state,
    zip: postal_code,
  };
};

export const toBillingDetails = ({
  address,
  cardholderName,
  email,
}: {
  address?: AddressInput;
  email?: string;
  cardholderName?: string;
  phone?: string;
}) => ({
  address: address
    ? {
        city: sanitise(address.city),
        country: address.countryCode,
        line1: sanitise(address.address1),
        line2: sanitise(address.address2),
        postal_code: sanitise(address.zip),
        state: sanitise(address.province),
      }
    : undefined,
  email,
  name: sanitise(cardholderName),
  phone: address ? sanitise(address.phone) : undefined,
});

export const toShippingOption = ({
  handle,
  price,
  title,
}: ShippingRate): PaymentRequestShippingOption => ({
  amount: toSmallestCurrencyUnit(price),
  detail: "", // ReCharge doesn't have a corresponding field, but it's required in the `PaymentRequestShippingOption`
  id: handle,
  label: title,
});

// Stripe expects all values to be provided in the smallest currency unit
// See https://stripe.com/docs/currencies#zero-decimal
export const toSmallestCurrencyUnit = (money: Money) => {
  const isZeroDecimal = [
    "BIF",
    "CLP",
    "DJF",
    "GNF",
    "JPY",
    "KMF",
    "KRW",
    "MGA",
    "PYG",
    "RWF",
    "UGX",
    "VND",
    "VUV",
    "XAF",
    "XOF",
    "XPF",
  ].includes(money.currencyCode);

  return Math.round(moneyFns.toFloat(money) * (isZeroDecimal ? 1 : 100));
};
