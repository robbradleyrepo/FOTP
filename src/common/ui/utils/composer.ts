import { evaluate, merge } from "./common";
import {
  IsomorphicStyleFn,
  ResponsiveCSSObject,
  Style,
  ThemedProps,
  ThemeEnhanced,
} from "./types";

const flatten = (args: Style[], theme: ThemeEnhanced): ResponsiveCSSObject =>
  args.reduce<ResponsiveCSSObject>(
    (acc, style) => merge(acc, evaluate(style, theme)),
    {}
  );

const isTheme = (arg: ThemeEnhanced | ThemedProps): arg is ThemedProps =>
  !!(arg.theme && Object.keys(arg.theme).length > 0);

export const s = (...styles: Style[]): IsomorphicStyleFn => (
  arg: ThemeEnhanced | ThemedProps
): any =>
  isTheme(arg)
    ? arg.theme.dynamicStyleFn(flatten(styles, arg.theme))
    : flatten(styles, arg);
