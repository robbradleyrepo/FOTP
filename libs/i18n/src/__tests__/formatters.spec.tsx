import { mocked } from "ts-jest/utils";

import { useLocale } from "../context";
import {
  formatCurrency,
  formatDateTime,
  useCurrencyFormatter,
  useDateTimeFormatter,
} from "../formatters";

jest.mock("../context", () => ({
  useLocale: jest.fn(() => ({
    locale: {
      countryCode: "US",
      langtag: "en-US",
      language: "en",
      region: "us",
      timeZone: "America/Los_Angeles",
    },
  })),
}));

const mockUseLocale = mocked(useLocale);

describe("formatCurrency", () => {
  it("should format the provided value using the provide locale and currency code", () => {
    expect(
      formatCurrency({ amount: "123", currencyCode: "USD", langtag: "en-US" })
    ).toBe("$123.00");
    expect(
      formatCurrency({ amount: "123", currencyCode: "GBP", langtag: "en-GB" })
    ).toBe("£123.00");
  });

  it("should format the provided value to the specified number of decimals", () => {
    expect(
      formatCurrency({
        amount: "123",
        currencyCode: "USD",
        fractionDigits: 0,
        langtag: "en-US",
      })
    ).toBe("$123");
    expect(
      formatCurrency({
        amount: "123",
        currencyCode: "GBP",
        fractionDigits: 0,
        langtag: "en-GB",
      })
    ).toBe("£123");
  });
});

describe("formatDateTime", () => {
  const dateString = "2020-02-01T12:00:00Z";

  it("should handle dates and parsable numbers and strings", () => {
    const timestamp = Date.parse(dateString);
    const date = new Date(dateString);

    expect(formatDateTime(dateString, "en-US")).toBe(
      formatDateTime(date, "en-US")
    );
    expect(formatDateTime(timestamp, "en-US")).toBe(
      formatDateTime(date, "en-US")
    );
  });

  it("should format the provided string or date using the provided locale and formatting options", () => {
    expect(formatDateTime(dateString.substring(0, 10), "en-US")).toBe(
      "2/1/2020"
    );
    expect(formatDateTime(dateString, "en-US")).toBe("2/1/2020");
    expect(formatDateTime(dateString, "en-GB")).toBe("01/02/2020");
    expect(
      formatDateTime(dateString, "en-US", {
        day: "numeric",
        month: "long",
        weekday: "long",
        year: "numeric",
      })
    ).toBe("Saturday, February 1, 2020");
    expect(
      formatDateTime(dateString.substring(0, 10), "en-US", {
        day: "numeric",
        month: "long",
        timeZone: "UTC",
        year: "numeric",
      })
    ).toEqual("February 1, 2020");
  });
});

describe("useCurrencyFormatter", () => {
  it("should format the provided amount, currency code and formatting options using the current locale", () => {
    [
      {
        amount: "123",
        currencyCode: "USD",
        langtag: "en-US",
        language: "en",
        region: "us",
      },
      {
        amount: "123",
        currencyCode: "GBP",
        langtag: "en-GB",
        language: "en",
        region: "gb",
      },
      {
        amount: "123",
        currencyCode: "USD",
        fractionDigits: 0,
        langtag: "en-US",
        language: "en",
        region: "us",
      },
      {
        amount: "123",
        currencyCode: "GBP",
        fractionDigits: 0,
        langtag: "en-GB",
        language: "en",
        region: "gb",
      },
    ].forEach(
      ({ amount, currencyCode, fractionDigits, langtag, language, region }) => {
        mockUseLocale.mockReturnValueOnce(({
          locale: {
            langtag,
            language,
            region,
          },
        } as unknown) as ReturnType<typeof useLocale>);

        const format = useCurrencyFormatter();

        expect(
          format({
            amount,
            currencyCode,
            fractionDigits,
          })
        ).toBe(
          formatCurrency({ amount, currencyCode, fractionDigits, langtag })
        );
      }
    );
  });

  it("should use the default formatting options if no explicit formatting options are provided", () => {
    const amount = "123";
    const currencyCode = "USD";
    const fractionDigits = 0;
    const langtag = "en-US";
    const language = "en";
    const region = "us";

    mockUseLocale.mockReturnValueOnce(({
      locale: {
        langtag,
        language,
        region,
      },
    } as unknown) as ReturnType<typeof useLocale>);

    const format = useCurrencyFormatter({ fractionDigits });

    expect(
      format({
        amount,
        currencyCode,
      })
    ).toBe(formatCurrency({ amount, currencyCode, fractionDigits, langtag }));

    expect(
      format({
        amount,
        currencyCode,
        fractionDigits: 2,
      })
    ).toBe(
      formatCurrency({ amount, currencyCode, fractionDigits: 2, langtag })
    );
  });
});

describe("useDateTimeFormatter", () => {
  const dateString = "2020-02-01T12:00:00Z";

  it("should format the provided date and formatting options using the current locale", () => {
    [
      { langtag: "en-US" },
      { langtag: "en-GB" },
      {
        langtag: "en-US",
        options: {
          day: "numeric",
          month: "long",
          weekday: "long",
          year: "numeric",
        },
      },
      {
        langtag: "en-GB",
        options: {
          day: "numeric",
          month: "long",
          weekday: "long",
          year: "numeric",
        },
      },
    ].forEach(({ langtag, options }) => {
      mockUseLocale.mockReturnValueOnce(({
        locale: { langtag },
      } as unknown) as ReturnType<typeof useLocale>);

      const format = useDateTimeFormatter();

      expect(format(dateString, options)).toBe(
        formatDateTime(dateString, langtag, options)
      );
    });
  });

  it("should use the default timezone if no timezone is provided", () => {
    const defaultOptions = {
      timeZone: "America/Los_Angeles",
    };
    const langtag = "en-US";
    const options = {
      timeZone: "Europe/London",
    };

    const format = useDateTimeFormatter(defaultOptions);

    expect(format(dateString)).toBe(
      formatDateTime(dateString, langtag, defaultOptions)
    );

    expect(format(dateString, {})).toBe(
      formatDateTime(dateString, langtag, defaultOptions)
    );

    expect(format(dateString, options)).toBe(
      formatDateTime(dateString, langtag, options)
    );
  });

  it("should combine the default and explicity formatting options", () => {
    const defaultOptions = {
      day: "numeric",
      hour: "numeric",
      minute: "numeric",
      month: "long",
      second: "numeric",
      weekday: "long",
      year: "numeric",
    };
    const options = { month: "short", weekday: "short", year: undefined };

    const format = useDateTimeFormatter(defaultOptions);

    expect(format(dateString)).toMatchSnapshot();

    expect(format(dateString, options)).toMatchSnapshot();
  });

  it("should fallback to UTC for invalid or unhandled timezones", () => {
    const format = useDateTimeFormatter();

    expect(format(dateString, { timeZone: "Foo/Bar" })).toMatchSnapshot();
  });
});
