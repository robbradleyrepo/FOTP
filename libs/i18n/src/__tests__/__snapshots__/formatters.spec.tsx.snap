// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useDateTimeFormatter should combine the default and explicity formatting options 1`] = `"Saturday, February 1, 2020, 4:00:00 AM"`;

exports[`useDateTimeFormatter should combine the default and explicity formatting options 2`] = `"Sat, Feb 1, 4:00:00 AM"`;

exports[`useDateTimeFormatter should fallback to UTC for invalid or unhandled timezones 1`] = `"2/1/2020, UTC"`;
